src/window.cpp

```diff
@@ -864,22 +864,17 @@ static void DispatchMouseWheelEvent(Window *w, NWidgetCore *nwid, int wheel)
 }
 
 /**
- * Generate repaint events for the visible part of window w within the rectangle.
- *
- * The function goes recursively upwards in the window stack, and splits the rectangle
- * into multiple pieces at the window edges, so obscured parts are not redrawn.
- *
- * @param w Window that needs to be repainted
- * @param left Left edge of the rectangle that should be repainted
- * @param top Top edge of the rectangle that should be repainted
- * @param right Right edge of the rectangle that should be repainted
- * @param bottom Bottom edge of the rectangle that should be repainted
- * @param gfx_dirty Whether to mark gfx dirty
+ * Helper class to hide the implementation details of \c DrawOverlappedWindow, below.
  */
-void DrawOverlappedWindow(Window *w, const int left, const int top, const int right, const int bottom, const bool gfx_dirty)
-{
-	std::function<void(Window const * const, int, int, int, int)> draw_overlapped;
-	draw_overlapped = [&w, &gfx_dirty, &draw_overlapped](Window const * const w_from, int left, int top, int right, int bottom){
+class DrawOverlappedWindowHelper {
+	Window *w; ///< The window being redrawn.
+	bool gfx_dirty; ///< The window is being redrawn within the context of \c DrawDirtyBlocks.
+
+public:
+	DrawOverlappedWindowHelper(Window *w, const bool gfx_dirty) : w(w), gfx_dirty(gfx_dirty) {}
+
+	void operator()(Window const * const w_from, int left, int top, int right, int bottom)
```

Comment:
I guess this is the sort of thing that @nielsmh had in mind. It still feels a bit ugly to me, IMHO; but if it's the best way to emulate nested functions, it's good enough for me.

Reply:
After playing around with the optimizer and disassembler, I've found that this construct actually does work. GCC has no problem inlining the construction and calling of the temporary object; and even unrolling a few iterations of recursive calls. Strange as it may seem, it looks like this is the way to do local functions in C++11 :open_mouth: